### ISSUE TITLE
Homebridge not starting after Update

### ISSUE THREAD

**Issue by chigy110** [homebridge/homebridge#3536]
### Analysis

After Updating Homebridge to v1.7.0 it's not starting anymore, could you please assist?

```
[1/9/2024, 12:23:07 PM] [HB Supervisor] Restarting Homebridge...
[1/9/2024, 12:23:07 PM] [HB Supervisor] Starting Homebridge with extra flags: -I -P /var/lib/homebridge/node_modules --strict-plugin-resolution
[1/9/2024, 12:23:07 PM] [HB Supervisor] Started Homebridge v1.7.0 with PID: 11896
[1/9/2024, 12:23:08 PM] Loaded config.json with 2 accessories and 7 platforms.
[1/9/2024, 12:23:08 PM] Loaded 28 cached accessories from cachedAccessories.
[1/9/2024, 12:23:09 PM] TypeError: Cannot read properties of undefined (reading 'getCharacteristic')
    at Accessory._sideloadServices (/homebridge/node_modules/homebridge/node_modules/hap-nodejs/dist/lib/Accessory.js:1800:13)
    at Function.Accessory.deserialize (/homebridge/node_modules/homebridge/node_modules/hap-nodejs/src/lib/Accessory.ts:2188:15)
    at Function.deserialize (/homebridge/node_modules/homebridge/src/platformAccessory.ts:172:33)
    at /homebridge/node_modules/homebridge/src/bridgeService.ts:240:34
    at Array.map (<anonymous>)
    at BridgeService.loadCachedPlatformAccessoriesFromDisk (/homebridge/node_modules/homebridge/src/bridgeService.ts:239:58)
    at Server.start (/homebridge/node_modules/homebridge/src/server.ts:170:5)
[1/9/2024, 12:23:09 PM] Got SIGTERM, shutting down Homebridge...
[1/9/2024, 12:23:09 PM] AssertionError [ERR_ASSERTION]: Cannot generate setupURI on an accessory that isn't published yet!
    at Bridge.Accessory.setupURI (/homebridge/node_modules/homebridge/node_modules/hap-nodejs/src/lib/Accessory.ts:994:11)
    at Server.setServerStatus (/homebridge/node_modules/homebridge/src/server.ts:155:45)
    at Server.teardown (/homebridge/node_modules/homebridge/src/server.ts:199:10)
    at signalHandler (/homebridge/node_modules/homebridge/src/cli.ts:93:12)
    at process.emit (node:events:514:28)
    at process.emit (/homebridge/node_modules/homebridge/node_modules/source-map-support/source-map-support.js:516:21)
[1/9/2024, 12:23:14 PM] [HB Supervisor] Homebridge Process Ended. Code: 143, Signal: null
[1/9/2024, 12:23:19 PM] [HB Supervisor] Restarting Homebridge...
[1/9/2024, 12:23:19 PM] [HB Supervisor] Starting Homebridge with extra flags: -I -P /var/lib/homebridge/node_modules --strict-plugin-resolution
[1/9/2024, 12:23:19 PM] [HB Supervisor] Started Homebridge v1.7.0 with PID: 11910
[1/9/2024, 12:23:19 PM] Loaded config.json with 2 accessories and 7 platforms.
[1/9/2024, 12:23:20 PM] Loaded 28 cached accessories from cachedAccessories.
[1/9/2024, 12:23:20 PM] TypeError: Cannot read properties of undefined (reading 'getCharacteristic')
    at Accessory._sideloadServices (/homebridge/node_modules/homebridge/node_modules/hap-nodejs/dist/lib/Accessory.js:1800:13)
    at Function.Accessory.deserialize (/homebridge/node_modules/homebridge/node_modules/hap-nodejs/src/lib/Accessory.ts:2188:15)
    at Function.deserialize (/homebridge/node_modules/homebridge/src/platformAccessory.ts:172:33)
    at /homebridge/node_modules/homebridge/src/bridgeService.ts:240:34
    at Array.map (<anonymous>)
    at BridgeService.loadCachedPlatformAccessoriesFromDisk (/homebridge/node_modules/homebridge/src/bridgeService.ts:239:58)
    at Server.start (/homebridge/node_modules/homebridge/src/server.ts:170:5)
[1/9/2024, 12:23:20 PM] Got SIGTERM, shutting down Homebridge...
[1/9/2024, 12:23:20 PM] AssertionError [ERR_ASSERTION]: Cannot generate setupURI on an accessory that isn't published yet!
    at Bridge.Accessory.setupURI (/homebridge/node_modules/homebridge/node_modules/hap-nodejs/src/lib/Accessory.ts:994:11)
    at Server.setServerStatus (/homebridge/node_modules/homebridge/src/server.ts:155:45)
    at Server.teardown (/homebridge/node_modules/homebridge/src/server.ts:199:10)
    at signalHandler (/homebridge/node_modules/homebridge/src/cli.ts:93:12)
    at process.emit (node:events:514:28)
    at process.emit (/homebridge/node_modules/homebridge/node_modules/source-map-support/source-map-support.js:516:21)
[1/9/2024, 12:23:25 PM] [HB Supervisor] Homebridge Process Ended. Code: 143, Signal: null
```

### Expected Behavior

Homebridge should start again

### Steps To Reproduce

-

### Logs

```markdown
[1/9/2024, 12:23:07 PM] [HB Supervisor] Restarting Homebridge...
[1/9/2024, 12:23:07 PM] [HB Supervisor] Starting Homebridge with extra flags: -I -P /var/lib/homebridge/node_modules --strict-plugin-resolution
[1/9/2024, 12:23:07 PM] [HB Supervisor] Started Homebridge v1.7.0 with PID: 11896
[1/9/2024, 12:23:08 PM] Loaded config.json with 2 accessories and 7 platforms.
[1/9/2024, 12:23:08 PM] Loaded 28 cached accessories from cachedAccessories.
[1/9/2024, 12:23:09 PM] TypeError: Cannot read properties of undefined (reading 'getCharacteristic')
    at Accessory._sideloadServices (/homebridge/node_modules/homebridge/node_modules/hap-nodejs/dist/lib/Accessory.js:1800:13)
    at Function.Accessory.deserialize (/homebridge/node_modules/homebridge/node_modules/hap-nodejs/src/lib/Accessory.ts:2188:15)
    at Function.deserialize (/homebridge/node_modules/homebridge/src/platformAccessory.ts:172:33)
    at /homebridge/node_modules/homebridge/src/bridgeService.ts:240:34
    at Array.map (<anonymous>)
    at BridgeService.loadCachedPlatformAccessoriesFromDisk (/homebridge/node_modules/homebridge/src/bridgeService.ts:239:58)
    at Server.start (/homebridge/node_modules/homebridge/src/server.ts:170:5)
[1/9/2024, 12:23:09 PM] Got SIGTERM, shutting down Homebridge...
[1/9/2024, 12:23:09 PM] AssertionError [ERR_ASSERTION]: Cannot generate setupURI on an accessory that isn't published yet!
    at Bridge.Accessory.setupURI (/homebridge/node_modules/homebridge/node_modules/hap-nodejs/src/lib/Accessory.ts:994:11)
    at Server.setServerStatus (/homebridge/node_modules/homebridge/src/server.ts:155:45)
    at Server.teardown (/homebridge/node_modules/homebridge/src/server.ts:199:10)
    at signalHandler (/homebridge/node_modules/homebridge/src/cli.ts:93:12)
    at process.emit (node:events:514:28)
    at process.emit (/homebridge/node_modules/homebridge/node_modules/source-map-support/source-map-support.js:516:21)
[1/9/2024, 12:23:14 PM] [HB Supervisor] Homebridge Process Ended. Code: 143, Signal: null
[1/9/2024, 12:23:19 PM] [HB Supervisor] Restarting Homebridge...
[1/9/2024, 12:23:19 PM] [HB Supervisor] Starting Homebridge with extra flags: -I -P /var/lib/homebridge/node_modules --strict-plugin-resolution
[1/9/2024, 12:23:19 PM] [HB Supervisor] Started Homebridge v1.7.0 with PID: 11910
[1/9/2024, 12:23:19 PM] Loaded config.json with 2 accessories and 7 platforms.
[1/9/2024, 12:23:20 PM] Loaded 28 cached accessories from cachedAccessories.
[1/9/2024, 12:23:20 PM] TypeError: Cannot read properties of undefined (reading 'getCharacteristic')
    at Accessory._sideloadServices (/homebridge/node_modules/homebridge/node_modules/hap-nodejs/dist/lib/Accessory.js:1800:13)
    at Function.Accessory.deserialize (/homebridge/node_modules/homebridge/node_modules/hap-nodejs/src/lib/Accessory.ts:2188:15)
    at Function.deserialize (/homebridge/node_modules/homebridge/src/platformAccessory.ts:172:33)
    at /homebridge/node_modules/homebridge/src/bridgeService.ts:240:34
    at Array.map (<anonymous>)
    at BridgeService.loadCachedPlatformAccessoriesFromDisk (/homebridge/node_modules/homebridge/src/bridgeService.ts:239:58)
    at Server.start (/homebridge/node_modules/homebridge/src/server.ts:170:5)
[1/9/2024, 12:23:20 PM] Got SIGTERM, shutting down Homebridge...
[1/9/2024, 12:23:20 PM] AssertionError [ERR_ASSERTION]: Cannot generate setupURI on an accessory that isn't published yet!
    at Bridge.Accessory.setupURI (/homebridge/node_modules/homebridge/node_modules/hap-nodejs/src/lib/Accessory.ts:994:11)
    at Server.setServerStatus (/homebridge/node_modules/homebridge/src/server.ts:155:45)
    at Server.teardown (/homebridge/node_modules/homebridge/src/server.ts:199:10)
    at signalHandler (/homebridge/node_modules/homebridge/src/cli.ts:93:12)
    at process.emit (node:events:514:28)
    at process.emit (/homebridge/node_modules/homebridge/node_modules/source-map-support/source-map-support.js:516:21)
[1/9/2024, 12:23:25 PM] [HB Supervisor] Homebridge Process Ended. Code: 143, Signal: null
```

### Configuration

```json
{
    "bridge": {
        "name": "Homebridge A296",
        "username": "",
        "port": 53784,
        "pin": ""
    },
    "accessories": [
        {
            "name": "Saugroboter",
            "ip": "192.168.0.59",
            "token": "",
            "waterBox": false,
            "silent": false,
            "pause": true,
            "pauseWord": "pause",
            "findMe": false,
            "findMeWord": "where are you",
            "dock": true,
            "cleanword": "cleaning",
            "roomTimeout": 0,
            "autoroom": false,
            "accessory": "XiaomiRoborockVacuum"
        },
        {
            "name": "Luftreiniger",
            "address": "192.168.0.35",
            "token": "",
            "enableAirQuality": true,
            "enableTemperature": true,
            "enableHumidity": true,
            "filterChangeThreshold": 5,
            "enableFanSpeedControl": false,
            "enableChildLockControl": false,
            "accessory": "XiaomiMiAirPurifier"
        }
    ],
    "platforms": [
        {
            "name": "Config",
            "port": 8581,
            "platform": "config"
        },
        {
            "name": "Hue",
            "anyOn": true,
            "effects": true,
            "hosts": [
                "192.168.0.4"
            ],
            "lights": true,
            "nativeHomeKitLights": true,
            "nativeHomeKitSensors": true,
            "resource": true,
            "platform": "Hue",
            "users": {
                "": ""
            }
        },
        {
            "platform": "AutomationSwitches",
            "switches": [
                {
                    "type": "switch",
                    "name": "Heizung Sommer Schaltung",
                    "stored": true,
                    "default": false
                },
                {
                    "type": "switch",
                    "name": "Olivia anwesend",
                    "stored": true,
                    "default": false
                },
                {
                    "type": "switch",
                    "name": "Marco anwesend",
                    "stored": true,
                    "default": false
                },
                {
                    "type": "switch",
                    "name": "EsszimmerSzene aktiv",
                    "stored": true,
                    "default": false
                },
                {
                    "type": "switch",
                    "name": "WohnzimmerSzene aktiv",
                    "stored": true,
                    "default": false
                },
                {
                    "type": "switch",
                    "name": "WohnEsszimmerSzene aktiv",
                    "stored": true,
                    "default": false
                }
            ]
        },
        {
            "name": "Nuki",
            "platform": "NB"
        },
        {
            "mqtt": {
                "base_topic": "zigbee2mqtt",
                "server": "mqtt://192.168.0.10:1883",
                "reject_unauthorized": false,
                "keepalive": 60,
                "version": 4,
                "disable_qos": false
            },
            "defaults": {
                "exclude": false
            },
            "exclude_grouped_devices": false,
            "platform": "zigbee2mqtt"
        },
        {
            "clientid": "",
            "simulator": false,
            "language": {
                "api": "de-DE"
            },
            "platform": "HomeConnect"
        },
        {
            "authCode": "032-104",
            "pollInterval": 60,
            "platform": "laundrify"
        }
    ],
    "disabledPlugins": [
        "homebridge-fhem"
    ]
}
```

### Environment

- OS: (Docker)  Ubuntu Focal Fossa (20.04.5 LTS)
- Node.js Version | v20.10.0
- HB-Version v1.7.0

### Process Supervisor

Docker (Mention image name in `Additional Context`)

### Additional Context

oznu/homebridge/

**Comment by chigy110** [comment]
had to delete cachedAccessories under /homebridge - now it's working again

**Comment by JonnyFPV** [comment]
@chigy110 : had the same, had to delete cache too, working again :-) thank you

**Comment by ventz** [comment]
Worked here too (had to re-do the room assignments, but not too bad)

Settings -> 
<img width="347" alt="image" src="https://github.com/homebridge/homebridge/assets/901168/d7a473af-7cae-445a-9679-94ba2c3381e4">

**Comment by movy** [comment]
This is a very bad error message (no clarity for end-user) and even worse solution is proposed here as removing ALL cached accessories will ruin lots of your Home app automations and you'd have to re-create them from scratch MANUALLY. Total insanity for larger homes with dozens of automations accumulated over the years.

Anyhow, a bit late for me, as my home setup is pretty much ruined by this upgrade, but for others: do NOT reset your accessories cache unless your setup is very simple. Instead, manually open your `/var/lib/homebridge/accessories/cachedAccessories` (location may vary, see https://github.com/homebridge/homebridge/wiki/How-To-Fix-Node.js-Install-Issues-On-Linux#homebridge-storage-folder),

then inside `cachedAccessories` look for an accessory that seems very empty (i.e. empty arrays, some `nulls` etc), example:

```json
{
        "category": 1,
        "context": {
            "device": {
                "definition": null,
                "disabled": false,
                "endpoints": {
                    "1": {
                        "bindings": [],
                        "clusters": {
                            "input": [],
                            "output": []
                        },
                        "configured_reportings": [],
                        "scenes": []
                    },
                    "2": {
                        "bindings": [],
                        "clusters": {
                            "input": [],
                            "output": []
                        },
                        "configured_reportings": [],
                        "scenes": []
                    },
                    "3": {
                        "bindings": [],
                        "clusters": {
                            "input": [],
                            "output": []
                        },
                        "configured_reportings": [],
                        "scenes": []
                    },
                    "4": {
                        "bindings": [],
                        "clusters": {
                            "input": [],
                            "output": []
                        },
                        "configured_reportings": [],
                        "scenes": []
                    },
                    "5": {
                        "bindings": [],
                        "clusters": {
                            "input": [],
                            "output": []
                        },
                        "configured_reportings": [],
                        "scenes": []
                    },
                    "6": {
                        "bindings": [],
                        "clusters": {
                            "input": [],
                            "output": []
                        },
                        "configured_reportings": [],
                        "scenes": []
                    },
                    "8": {
                        "bindings": [],
                        "clusters": {
                            "input": [],
                            "output": []
                        },
                        "configured_reportings": [],
                        "scenes": []
                    },
                    "10": {
                        "bindings": [],
                        "clusters": {
                            "input": [],
                            "output": []
                        },
                        "configured_reportings": [],
                        "scenes": []
                    },
                    "11": {
                        "bindings": [],
                        "clusters": {
                            "input": [
                                "ssIasAce"
                            ],
                            "output": [
                                "ssIasWd",
                                "ssIasZone"
                            ]
                        },
                        "configured_reportings": [],
                        "scenes": []
                    },
                    "12": {
                        "bindings": [],
                        "clusters": {
                            "input": [],
                            "output": []
                        },
                        "configured_reportings": [],
                        "scenes": []
                    },
                    "13": {
                        "bindings": [],
                        "clusters": {
                            "input": [
                                "genOta"
                            ],
                            "output": []
                        },
                        "configured_reportings": [],
                        "scenes": []
                    },
                    "47": {
                        "bindings": [],
                        "clusters": {
                            "input": [],
                            "output": []
                        },
                        "configured_reportings": [],
                        "scenes": []
                    },
                    "110": {
                        "bindings": [],
                        "clusters": {
                            "input": [],
                            "output": []
                        },
                        "configured_reportings": [],
                        "scenes": []
                    },
                    "242": {
                        "bindings": [],
                        "clusters": {
                            "input": [],
                            "output": []
                        },
                        "configured_reportings": [],
                        "scenes": []
                    }
                },
                "friendly_name": "Coordinator",
                "ieee_address": "0x00124b00258d3b72",
                "interview_completed": true,
                "interviewing": false,
                "network_address": 0,
                "supported": true,
                "type": "Coordinator"
            }
        },
        "displayName": "Coordinator",
        "platform": "zigbee2mqtt",
        "plugin": "homebridge-z2m",
        "services": [],
        "UUID": "ea330e77-e868-431d-a14c-9acaca3e32c6"
    }
    
```

Now you can go back to Homebridge UI and in settings instead of removing all cached accessories, remove this **single** one and see if it helps. There might be several of these, so remove them one by one to find the culprit.

Hope it saves someone a day of restoring your home to a working condition.

To devs: this error message is very generic and essentially useless as it lacks any details about the failed accessory. It may be a good idea to improve parsing function so the UUID of a failed device ends up in catch block so user can somehow act upon it, or even remove such device from cache automatically.

**Comment by donavanbecker** [comment]
Child bridges would also help in this situation as only the plugin (child bridge) that is causing the issue would be in the restart loop and all others would still be functioning.

https://github.com/homebridge/homebridge/wiki/Child-Bridges

**Comment by movy** [comment]
3 months later update. Today all of a sudden my a/c kept running when a window was open, and after a brief investigation turned out my homebridge is in looping reboot with the same error:
`Homebridge Process Ended. Code: 143, Signal: null`

I did not touch my HB instance since the clusterfudge caused by the most recent upgrade (beginning of this issue), so this was totally out of the blue. Not willing to re-setup my whole smart home again, I carefully investigated differences between current `/var/lib/homebridge` dir contents and the most recent backup in `/var/lib/homebridge/backups/instance-backups` (this is seriously a lifesaver btw, thanks for implementing it out of the box). 

Almost no significant changes in any files (expected), but what saved my setup was `IdentifierCache.0E2CDF27791E.json` -- I just replaced the current one with the most recent backed up one and now my HB is back alive again w/o need to re-add everything! 

![image](https://github.com/homebridge/homebridge/assets/4336560/d14f9531-6707-4b95-8fee-65229cdcff43)


What really bothers me: since this crash was not caused by any config changes or plugin updates,  this can happen at any time when home is unattended and if I don't notice it in time, everything can turn into pumpkin once the working backup is overwritten... not cool at all.

TL;DR: fetch `/storage/persist/IdentifierCache.*.json` from your most recent backup located in /var/lib/homebridge/backups/instance-backups and overwrite the current IdentifierCache with the recovered file.

p.s. a funny coincidence: this issue was marked stale while I was typing this response. Not so fast, robots!

**Comment by NorthernMan54** [comment]
FYI - The stale is just doing its thing, just ignore it

Any thoughts on how to recreate the issue, so we can improve the code base ?

**Comment by movy** [comment]
Man, I wish I knew what happened. Both IdentifierCache are attached, have a look, maybe something obvious pops out. As many mentioned here and elsewhere online, there's 0 information related to the crash in HB logs despite all the verbosity. It just happens, w/o any human involvement it seems.

[this_works_IdentifierCache.0E2CDF27791E.json](https://github.com/homebridge/homebridge/files/14695570/this_works_IdentifierCache.0E2CDF27791E.json)
[this_crashes_IdentifierCache.0E2CDF27791E.json](https://github.com/homebridge/homebridge/files/14695571/this_crashes_IdentifierCache.0E2CDF27791E.json)

I could not see any other significant difference in working backup vs broken current configs... that's all I know for now.

**Comment by 1004PTLN** [comment]
I am not a programmer so take this with a big grain of salt. I compared the two files and in the working file, the `nextAID` value is `326`, and there is no entry for `5c7554d8-f473-432c-b4c0-6bf68482b2e2`.

In the crashing file, the `nextAID` value is `327`, and the entry for `5c7554d8-f473-432c-b4c0-6bf68482b2e2` is present with a value of `326`.

Could this discrepancy be due to a synchronization issue or a state change that was captured in the crashing file but not in the working file?

**Comment by NorthernMan54** [comment]
@movy @1004PTLN Thanks for the file, going to put it into my test environment and look at defensive code around this.  Am thinking a synchronization issue between the various Homebridge state files, and the corruption occurred by homebridge exiting while saving.

@Supereg Am thinking this may be a HAP-NodeJS issue?

**Comment by bauer-andreas** [comment]
Have we verified that this issue only occurs on 1.7.0? Cause looking at the version diff there weren’t any major changes in that release that could cause any issues like these https://github.com/homebridge/homebridge/compare/v1.6.1...v1.7.0.
HAP-NodeJS version also didn’t change in that release.

Were there any changes in the node runtime itself that could cause that?

Otherwise, improving the parsing of these files seems like a good first step. At the same time we should investigate why the accessories end up being saved to disk incompletely.
Seems like it’s not just cached accessories (which is only used with homebridge) but also files like the identifier cache etc? so just weird accessory instances floating around? 


EDIT: Looking at that reply https://github.com/homebridge/homebridge/issues/3536#issuecomment-1899833277, the data that is considered „empty“ is all coming from the zigbee2mqtt platform plugin.

**Comment by bauer-andreas** [comment]
For this case https://github.com/homebridge/homebridge/issues/3536#issuecomment-2012078868, can we get the full log with the actual crash in it? Not just the terminating message. Additionally, the list of configured plugins would also be helpful.

**Comment by movy** [comment]
I use latest homebridge-z2m 1.9.3, plugins and cachedAccessories are attached. I removed all api keys etc, so might be slightly incomplete. 
As for the crash log, it was quite bloated due to several hours of looping, with all the plugins init logging, but then sudden crash with `Homebridge Process Ended. Code: 143, Signal: null`, so did not seem very useful and I deleted it. Next time will save it.

[plugins.json](https://github.com/homebridge/homebridge/files/14733226/plugins.json)

[cachedAccessories.json](https://github.com/homebridge/homebridge/files/14733238/cachedAccessories.json)

**Comment by bauer-andreas** [comment]
On thing to note, Code 143 is usually SIGTERM on Linux. So this is not a crash that is happening from within homebridge, but an external system killing homebridge process (e.g., could also be the UI, e.g., shutting down homebridge).
Definitely a different issue than the one originally reported here.

Thanks for all the additional information.

**Comment by movy** [comment]
Yes, totally agree, the cause of my recent issue is different, but the crash itself is also marked with code 143 (and nothing else), so googling it leads here.
This time there was less clarity about causes, as there was no `AssertionError` or `TypeError` preceding the crash. Worth opening another issue, to make it more prominent on Google, as I saw a number of similar posts on Reddit as well.

**Comment by ACQ1505** [comment]
I have the same problem today

no modification in HB just an error 143 with got sigterm

json update

always the same

a lot of red in the tuyaopenAI plugin

**Comment by ACQ1505** [comment]
it came back like it left

reboot linux
[homebridge.log(3).txt](https://github.com/homebridge/homebridge/files/14795361/homebridge.log.3.txt)

**Comment by BKrnk1337** [comment]
I stumbled across the same problem mention by @movy - till now I did not start to (at least) try to fix it.

Does anyone have ideas what is the smartest way now?
